pkg/kubecost/summaryallocation.go

```diff
@@ -0,0 +1,1155 @@
+// is usage or cost, then efficiency is 100%.
+func (sa *SummaryAllocation) CPUEfficiency() float64 {
+	if sa.CPUCoreRequestAverage > 0 {
+		return sa.CPUCoreUsageAverage / sa.CPUCoreRequestAverage
+	}
+
+	if sa.CPUCoreUsageAverage == 0.0 || sa.CPUCost == 0.0 {
+		return 0.0
+	}
+
+	return 1.0
+}
+
+func (sa *SummaryAllocation) generateKey(aggregateBy []string, labelConfig *LabelConfig) string {
+	if sa == nil {
+		return ""
+	}
+
+	return sa.Properties.GenerateKey(aggregateBy, labelConfig)
+}
+
+// IsExternal is true if the given Allocation represents external costs.
+func (sa *SummaryAllocation) IsExternal() bool {
+	return strings.Contains(sa.Name, ExternalSuffix)
+}
+
+// IsIdle is true if the given Allocation represents idle costs.
+func (sa *SummaryAllocation) IsIdle() bool {
+	return strings.Contains(sa.Name, IdleSuffix)
+}
+
+// IsUnallocated is true if the given Allocation represents unallocated costs.
+func (sa *SummaryAllocation) IsUnallocated() bool {
+	return strings.Contains(sa.Name, UnallocatedSuffix)
+}
+
+// IsUnmounted is true if the given Allocation represents unmounted volume costs.
+func (sa *SummaryAllocation) IsUnmounted() bool {
+	return strings.Contains(sa.Name, UnmountedSuffix)
+}
+
+// Minutes returns the number of minutes the SummaryAllocation represents, as
+// defined by the difference between the end and start times.
+func (sa *SummaryAllocation) Minutes() float64 {
+	return sa.End.Sub(sa.Start).Minutes()
+}
+
+// RAMEfficiency is the ratio of usage to request. If there is no request and
+// no usage or cost, then efficiency is zero. If there is no request, but there
+// is usage or cost, then efficiency is 100%.
+func (sa *SummaryAllocation) RAMEfficiency() float64 {
+	if sa.RAMBytesRequestAverage > 0 {
+		return sa.RAMBytesUsageAverage / sa.RAMBytesRequestAverage
+	}
+
+	if sa.RAMBytesUsageAverage == 0.0 || sa.RAMCost == 0.0 {
+		return 0.0
+	}
+
+	return 1.0
+}
+
+// TotalCost is the total cost of the SummaryAllocation
+func (sa *SummaryAllocation) TotalCost() float64 {
+	return sa.CPUCost + sa.GPUCost + sa.RAMCost + sa.PVCost + sa.NetworkCost + sa.LoadBalancerCost + sa.SharedCost + sa.ExternalCost
+}
+
+// TotalEfficiency is the cost-weighted average of CPU and RAM efficiency. If
+// there is no cost at all, then efficiency is zero.
+func (sa *SummaryAllocation) TotalEfficiency() float64 {
+	if sa.RAMCost+sa.CPUCost > 0 {
+		ramCostEff := sa.RAMEfficiency() * sa.RAMCost
+		cpuCostEff := sa.CPUEfficiency() * sa.CPUCost
+		return (ramCostEff + cpuCostEff) / (sa.CPUCost + sa.RAMCost)
+	}
+
+	return 0.0
+}
+
+// SummaryAllocationSet stores a set of SummaryAllocations, each with a unique
+// name, that share a window. An AllocationSet is mutable, so treat it like a
+// threadsafe map.
+type SummaryAllocationSet struct {
+	sync.RWMutex
+	externalKeys       map[string]bool
+	idleKeys           map[string]bool
+	SummaryAllocations map[string]*SummaryAllocation `json:"allocations"`
```

Comment:
My guess is that this unfortunate circumstance is because of the auto-serialization approach of `encoding/json` that requires exported fields (yay, Golang).

Reply:
That's correct. And there's a slightly bigger story. Basically, when I designed `AllocationSet` the first time, I was sensitive to the idea that it might end up being a core type, which others (perhaps even outside the company) might want to use for a very wide range of purposes. So safety seemed high-priority. Now, with `SummaryAllocation`, speed is the priority, and we've learned lessons about serialization. (Yay, Go, indeed.)

So this might be a little controversial. But it's also still kind of fair to say "treat it like a threadsafe map" haha. Like—use the methods! Don't just access the map! If you do, you're mis-using the type. But also, don't make me write "Each" and "Get" because you can just RWLock and iterate or access by key. I dunno. Maybe with this freedom, future code writers will shoot themselves in the foot. I hope not :shrug:

Reply:
:+1: :+1: I think this is excellent reasoning. Abusers of performance-focused code, beware! :grin: